ports/rp2/machine_uart.c

```diff
@@ -192,14 +305,28 @@ STATIC mp_obj_t machine_uart_make_new(const mp_obj_type_t *type, size_t n_args,
         if (self->invert & UART_INVERT_TX) {
             gpio_set_outover(self->tx, GPIO_OVERRIDE_INVERT);
         }
+        // allocate the buffers
+        ringbuf_alloc(&(self->read_buffer), rxbuf_len + 1);
+        MP_STATE_PORT(rp2_uart_rx_buffer[uart_id]) = self->read_buffer.buf;
+
+        ringbuf_alloc(&(self->write_buffer), txbuf_len + 1);
+        MP_STATE_PORT(rp2_uart_tx_buffer[uart_id]) = self->write_buffer.buf;
+
+        // set the irq handler
+        irq_set_exclusive_handler(self->irq_nr, self->irq_handler);
+        irq_set_enabled(self->irq_nr, true);
+        // enable the uart irq. this macro sets the rx irq level to 4
+        uart_set_irq_enables(self->uart, true, true);
     }
 
     return MP_OBJ_FROM_PTR(self);
 }
 
 STATIC mp_obj_t machine_uart_any(mp_obj_t self_in) {
     machine_uart_obj_t *self = MP_OBJ_TO_PTR(self_in);
-    return MP_OBJ_NEW_SMALL_INT(uart_is_readable(self->uart));
+    // get all bytes from the fifo first
+    uart_drain_rx_fifo(self, true);
```

Comment:
I don't think there needs to be any locking here.  The ringbuf is implemented so that it can be simultaneously read and written without any race conditions.

Reply:
I was not sure about the locking. I added that since during testing I had error during reading from the uart while data was sent to it. These errors went ways after adding the lock. The decision of full vs empty is done by comparing the read and write index. So reading and writing are not independent.

Reply:
When I disable the locking, the errors are back in the receive test.

Reply:
The concurrency is not between read and write of the ringbuf, but between two instances writing to the rinfbuf, a) the ISR, b) the polling loop in uart_read() and the call in uart_any(). The latter ones are needed because the ISR is not called until 4 items are in the FIFO.

Reply:
Yes, agreed.